File: java/0005_Longest_Palindromic_Substring.java
class Solution {
    private int len;
    private char[] chars;
    public String longestPalindrome(String s) {
        String answer = "";
        len = s.length();
        chars = s.toCharArray();

        for (int middle = 0; middle < len; middle++) {
            String oddCase = getPalindrome(s, middle, middle);
            String evenCase = getPalindrome(s, middle, middle + 1);

            answer = (oddCase.length() > answer.length()) ? oddCase : answer;
            answer = (evenCase.length() > answer.length()) ? evenCase : answer;
        }
        return answer;
    }

    private String getPalindrome(String s, int left, int right) {
        for (; 0 <= left && right < len; left--, right++) {
            if (chars[left] != chars[right]) break;
        }
        return s.substring(left + 1, right);
    }
}

File: kotlin/0005_Longest_Palindromic_Substring.kt
class Solution {
    fun longestPalindrome(s: String): String {
        var answer = ""
        for (middle in s.indices) {
            val oddCase = s.getPalindrome(middle, middle)
            val evenCase = s.getPalindrome(middle, middle + 1)
            answer = if (oddCase.length > answer.length) oddCase else answer
            answer = if (evenCase.length > answer.length) evenCase else answer
        }
        return answer
    }

    private fun String.getPalindrome(l: Int, r: Int): String {
        var left = l
        var right = r
        while (0 <= left && right < length) {
            if (this[left] != this[right]) break
            left--
            right++
        }
        return substring(left + 1, right)
    }
}
